pdns/opensslsigners.cc

```diff
@@ -5,9 +5,306 @@
+  SHA_CTX ctx;
+
+  int res = SHA1_Init(&ctx);
+
+  if (res != 1) {
+    throw runtime_error(getName()+" failed to init hash context for generating the public key hash");
+  }
+
+  int len = BN_bn2bin(d_key->e, tmp);
+  res = SHA1_Update(&ctx, tmp, len);
+  if (res != 1) {
+    throw runtime_error(getName()+" failed to update hash context for generating the public key hash");
+  }
+
+  len = BN_bn2bin(d_key->n, tmp);
+  res = SHA1_Update(&ctx, tmp, len);
+  if (res != 1) {
+    throw runtime_error(getName()+" failed to update hash context for generating the public key hash");
+  }
+
+  res = SHA1_Final(hash, &ctx);
+  if (res != 1) {
+    throw runtime_error(getName()+" failed to finish hash context for generating the public key hash");
+  }
+
+  return string((char*) hash, sizeof(hash));
+}
+
+
+std::string OpenSSLRSADNSCryptoKeyEngine::getPublicKeyString() const
+{
+  string keystring;
+  unsigned char tmp[std::max(BN_num_bytes(d_key->e), BN_num_bytes(d_key->n))];
+
+  int len = BN_bn2bin(d_key->e, tmp);
+  if (len < 255) {
+    keystring.assign(1, (char) (unsigned int) len);
+  } else {
+    keystring.assign(1, 0);
+    uint16_t tempLen = len;
+    tempLen = htons(tempLen);
+    keystring.append((char*)&tempLen, 2);
+  }
+  keystring.append((char *) tmp, len);
+
+  len = BN_bn2bin(d_key->n, tmp);
+  keystring.append((char *) tmp, len);
+
+  return keystring;
+}
+
+
+void OpenSSLRSADNSCryptoKeyEngine::fromISCMap(DNSKEYRecordContent& drc, std::map<std::string, std::string>& stormap)
+{
+  string sline;
+  string key,value;
+  typedef map<string, BIGNUM**> places_t;
+  places_t places;
+
+  places["Modulus"]=&d_key->n;
+  places["PublicExponent"]=&d_key->e;
+  places["PrivateExponent"]=&d_key->d;
+  places["Prime1"]=&d_key->p;
+  places["Prime2"]=&d_key->q;
+  places["Exponent1"]=&d_key->dmp1;
+  places["Exponent2"]=&d_key->dmq1;
+  places["Coefficient"]=&d_key->iqmp;
+
+  drc.d_algorithm = pdns_stou(stormap["algorithm"]);
+
+  string raw;
+  for(const places_t::value_type& val :  places) {
+    raw=stormap[toLower(val.first)];
+    *val.second = BN_bin2bn((unsigned char*) raw.c_str(), raw.length(), NULL);
+    if (!*val.second) {
+      throw runtime_error(getName()+" error loading " + val.first);
+    }
+  }
+
+  if (drc.d_algorithm != d_algorithm) {
+    throw runtime_error(getName()+" tried to feed an algorithm "+std::to_string(drc.d_algorithm)+" to a "+std::to_string(d_algorithm)+" key");
+  }
+
+  int ret = RSA_check_key(d_key);
+  if (ret != 1) {
+    throw runtime_error(getName()+" invalid public key");
+  }
+}
+
+void OpenSSLRSADNSCryptoKeyEngine::fromPublicKeyString(const std::string& input)
+{
+  string exponent, modulus;
+  const unsigned char* raw = (const unsigned char*)input.c_str();
+
+  if (raw[0] != 0) {
+    exponent = input.substr(1, raw[0]);
+    modulus = input.substr(raw[0]+1);
+  } else {
+    exponent = input.substr(3, raw[1]*0xff + raw[2]);
+    modulus = input.substr(3+ raw[1]*0xff + raw[2]);
+  }
+
```

Comment:
possibility for reading out of bounds and an uncaught out_of_range exception from substr() if input  is not big enough